aten/src/ATen/native/cpu/ScatterGatherKernel.cpp

```diff
@@ -275,7 +297,158 @@ void scatter_add_cpu_kernel(Tensor& self, int64_t dim, const Tensor& index, cons
+  public:
+    ReduceFunctor() {};
+    template <typename T, typename func_t>
+    void operator() (T* self_data, const T* src_data,
+                     func_t op) {
+      op(self_data, src_data);
+    }
+  };
+
+  ReduceFunctor reduce_fn;
+
+  auto reduce_sum = [](auto * self_data, const auto * src_data) {
+                      *self_data += *src_data;
+                    };
+
+  auto reduce_subtract = [](auto * self_data, const auto * src_data) {
+                           *self_data -= *src_data;
+                         };
+
+  auto reduce_multiply = [](auto * self_data, const auto * src_data) {
+                           *self_data *= *src_data;
+                         };
+
+  auto reduce_divide = [](auto * self_data, const auto * src_data) {
+                         std::cout << "self_data: " << *self_data << std::endl;
+                         std::cout << "src data: " << *src_data << std::endl;
+                         *self_data /= *src_data;
+                       };
+
+template <typename func_t>
+void cpu_scatter_gather_reduce_kernel(
+  Tensor& self, int64_t dim,
+  const Tensor& index, const Tensor& src,
+  const std::string& method_name,
+  const func_t& f,
+  const std::string reduce,
+  bool serial_exec = true
+) {
+  auto index_sizes = ensure_nonempty_vec(index.sizes().vec());
+  auto index_strides = ensure_nonempty_vec(index.strides().vec());
+
+  // `dim` is traversed in a kernel function `f`,
+  // that is why index.stride(dim) = 0 and index.size(dim) = 1.
+  // Also, index.size(dim) = 1 makes sure that TensorIterator.DimCounter
+  // has the following form : (i_1,..., i_{dim-1}, 0, i_{dim+1},...,i_n).
+  index_sizes[dim] = 1;
+  index_strides[dim] = 0;
+
+  // set self.shape = src.shape = index.shape,
+  // this defines the number of elements to iterate over,
+  // and set self.stride(dim) = src.stride(dim) = 0,
+  // because `dim` is traversed in a kernel function `f`.
+  auto self_restrided = restride_dim(self, dim, index_sizes);
+  auto index_restrided = index.as_strided(index_sizes, index_strides);
+  auto src_restrided = restride_dim(src, dim, index_sizes);
+
+  auto iter = TensorIterator();
+  iter.dont_compute_common_dtype();
+  iter.dont_resize_outputs();
+  iter.add_output(self_restrided);
+  iter.add_input(src_restrided, src.device(), src.scalar_type());
+  iter.add_input(index_restrided);
+  iter.build();
+
+  auto self_dim_stride = ensure_nonempty_stride(self, dim);
+  auto index_dim_stride = ensure_nonempty_stride(index, dim);
+  auto src_dim_stride = ensure_nonempty_stride(src, dim);
+
+  AT_DISPATCH_ALL_TYPES_AND2(
+    ScalarType::Bool, ScalarType::Half, iter.dtype(),
+    method_name, [&] {
+      using reduce_func_t = std::function<void(scalar_t*, const scalar_t*)>;
+      std::map<const std::string, reduce_func_t> reduce_funcs = {
+          {"sum", reduce_sum},
+          {"subtract", reduce_subtract},
+          {"multiply", reduce_multiply},
+          {"divide", reduce_divide}
+      };
```

Comment:
Why `map`?

Reply:
Also, maybe it is better so specialize the reduction operations here? It is a matter of taste, but I personally try to avoid any namespace clutter.

Reply:
I am also thinking is it possible to extend the reduce functions to accept parameters? Like, for example, powers...

Reply:
`unordered_map` does not have initialization features supported on the compiler I'm using (GCC 7.5.0). I'll change it to using `[]=` for initialization.